### PR TITLE
Fetch element by xpath

### PR DESCRIPTION
**Pull request by mdrokz**
# Changes

https://github.com/mdrokz/rust-headless-chrome/commit/82c4bb9664d96db7e20ef6165a0c6181898fa44d - implemented feature to find elements by xpath and created a generic function that takes an enum to find element by either css selector or xpath query.

### CODE REVIEW

File: src/browser/tab/element/mod.rs [hunk 1]
```diff
@@ -337,38 +341,65 @@ impl<'a> Element<'a> {
...
+            }) {
+            Ok(e) => return Ok(e),
+            Err(_) => {
+                let mut p = Point { x: 0.0, y: 0.0 };
+
+                util::Wait::with_sleep(Duration::from_secs(1)).run_until(|| {
+                    let r = self.call_js_fn(
+                        r#"
+                    function() {
+                        let v = document.getElementsByClassName("ShowAllChapters")[0];
```
Comment (atroche) [inline src/browser/tab/element/mod.rs:365]: ShowAllChapters?

File: src/browser/tab/mod.rs [hunk 2]
```diff
@@ -382,31 +400,31 @@ impl<'a> Tab {
         sleep(Duration::from_millis(scaled_millis));
     }
 
-    pub fn wait_for_element(&self, selector: &str) -> Fallible<Element<'_>> {
+    pub fn wait_for_element(&self, selector: Selector) -> Fallible<Element<'_>> {
```
Comment (atroche) [inline src/browser/tab/mod.rs:403]: It'd be nice to not have to break backwards compatibility here. Could you add a separate method for xpath, [like how Puppeteer has a separate method](https://pptr.dev/#?product=Puppeteer&version=v8.0.0&show=api-pagewaitforxpathxpath-options)?

Comment (atroche) [reply src/browser/tab/mod.rs:403]: Or use `Into` and make the default behavour for `"#asdfad".into()` be that it becomes a CSS selector?

File: src/util.rs [hunk 3]
```diff
@@ -70,6 +93,20 @@ impl Wait {
         }
     }
 
+    pub fn run_until<F>(&self, predicate: F)
```
Comment (atroche) [inline src/util.rs:96]: why do you need this method? in the case where you've used it, you can have the predicate return `Some(())` or `None` instead of true / false, right?

File: src/util.rs [hunk 4]
```diff
@@ -70,6 +93,20 @@ impl Wait {
...
+    where
+        F: FnMut() -> bool,
+    {
+        let mut predicate = predicate;
+
+        loop {
+            if predicate() {
```
Comment (atroche) [inline src/util.rs:103]: without using the timeout this could hang forever

File: src/browser/tab/mod.rs [hunk 5]
```diff
@@ -58,6 +58,12 @@ pub type ResponseHandler = Box<
     + Sync,
 >;
 
+pub struct SyncSendEvent(pub Arc<Tab>, pub Box<dyn Fn(&Event, &Tab)>);
```
Comment (atroche) [inline src/browser/tab/mod.rs:61]: is this so that the event listener can have a reference to the `Tab`? why can't you just clone the tab and `move` it into the event listener closure?

**Review by atroche — CHANGES_REQUESTED** [review]
In summary:

1. I don't think we should change the type of `SyncSendEvent` unless we really need to.
2. Let's keep `find_element` and its friends based on CSS selectors by default, to keep consistency with Puppeteer and also to keep the most common use case (which is definitely CSS selectors) ergonomic
3. The extra `run_until` method seems unnecessary and also dangerous (without timeout)

File: src/browser/tab/element/mod.rs (hunk 1 above)
Comment (mdrokz) [reply src/browser/tab/element/mod.rs:365]: Sorrr this is embarrasing, i forgot to remove this

File: src/browser/tab/mod.rs (hunk 2 above)
Comment (mdrokz) [reply src/browser/tab/mod.rs:403]: Originally i had a seperate method,but then i decided to follow what fantoccini did and made a function that accepted a enum, i will make a seperate method then,

File: src/browser/tab/mod.rs (hunk 5 above)
Comment (mdrokz) [reply src/browser/tab/mod.rs:61]: yeah i previously tried to move the tab into the closure. but that wasnt working, so i changed the SyncSendEvent to accept the tab struct

File: src/util.rs (hunk 3 above)
Comment (mdrokz) [reply src/util.rs:96]: i use this method to constantly check for midpoints after failure. when the dom doesnt load sometimes the midpoint extraction fails, i previously tried using a timeout, i will change the code to use a custom timeout.

File: src/browser/tab/mod.rs (hunk 5 above)
Comment (atroche) [reply src/browser/tab/mod.rs:61]: the reason i'm pushing back is that it'll break code that uses the current API, and I think you can get it working with moving the tab (maybe you need to make a new clone and move that?)

File: src/util.rs (hunk 3 above)
Comment (atroche) [reply src/util.rs:96]: but how come you can't just use the existing `until` method?

File: src/browser/tab/mod.rs (hunk 5 above)
Comment (mdrokz) [reply src/browser/tab/mod.rs:61]: Yeah i understand, this is my first time contributing, so im a bit inexperienced, i will try doing that. i need some guidance 😅

File: src/util.rs (hunk 3 above)
Comment (mdrokz) [reply src/util.rs:96]: i did use that before but it required me to return a result,which i felt unnecessary, since i could return true or false.

File: src/browser/tab/mod.rs (hunk 5 above)
Comment (atroche) [reply src/browser/tab/mod.rs:61]: no worries, thanks for taking the time to respond to my feedback 🙂

File: src/util.rs (hunk 4 above)
Comment (mdrokz) [reply src/util.rs:103]: i wanted to make this method for specific rare use cases such as on a slow internet the dom can render very late. but your right

File: src/browser/tab/mod.rs (hunk 2 above)
Comment (mdrokz) [reply src/browser/tab/mod.rs:403]: > Or use `Into` and make the default behavour for `"#asdfad".into()` be that it becomes a CSS selector?

this seems like a nice idea,  i will try doing that

File: src/browser/tab/mod.rs (hunk 5 above)
Comment (mdrokz) [reply src/browser/tab/mod.rs:61]: > no worries, thanks for taking the time to respond to my feedback

let me try again with cloning the tab and moving it in the closure.

File: src/browser/tab/element/mod.rs (hunk 1 above)
Comment (mdrokz) [reply src/browser/tab/element/mod.rs:365]: > Sorry this is embarrasing, i forgot to remove this

i was testing it with a element and forgot to remove it in my fork

File: src/util.rs (hunk 3 above)
Comment (atroche) [reply src/util.rs:96]: In your example you do:

```
p = v;
true
```

where p is a mut variable defined above the call to `run_until`. but you could instead do `p = until(.....)` and have the until return `Some(v)`, which will become `p`.

Comment (mdrokz) [reply src/util.rs:96]: > In your example you do:
> 
> ```
> p = v;
> true
> ```
> 
> where p is a mut variable defined above the call to `run_until`. but you could instead do `p = until(.....)` and have the until return `Some(v)`, which will become `p`.

I see, that definitely seems a better way.

File: src/browser/tab/mod.rs (hunk 2 above)
Comment (atroche) [reply src/browser/tab/mod.rs:403]: cool, and if you call `into()` inside `find_element` then people can still call `find_element('#example')` (i.e. it won't change the public API)

Comment (mdrokz) [reply src/browser/tab/mod.rs:403]: > cool, and if you call `into()` inside `find_element` then people can still call `find_element('#example')` (i.e. it won't change the public API)

Got it

File: src/browser/tab/mod.rs (hunk 5 above)
Comment (mdrokz) [reply src/browser/tab/mod.rs]: I tried again with cloning the tab and it was working, previously i didnt clone it properly.

**Comment by mdrokz** [comment]
> In summary:
> 
> 1. I don't think we should change the type of `SyncSendEvent` unless we really need to.
> 2. Let's keep `find_element` and its friends based on CSS selectors by default, to keep consistency with Puppeteer and also to keep the most common use case (which is definitely CSS selectors) ergonomic
> 3. The extra `run_until` method seems unnecessary and also dangerous (without timeout)

I fixed all of the issues you have mentioned in this commit - https://github.com/atroche/rust-headless-chrome/pull/239/commits/b68cfa53051d625c5fb773962688cc200627b31f

**Comment by atroche** [comment]
@mdrokz nice one!

With [this](https://github.com/atroche/rust-headless-chrome/commit/b68cfa53051d625c5fb773962688cc200627b31f#diff-d91ca1f76b204ad2cf6d1781c3fbbd7559d0dd60b553dac372772308c14db558R79-R90), I think there are ways that can go wrong (e.g. if you use an XPath expression that contains a `.` or a `#` in it).

Instead, could you make strings always convert to a `Selector::CSS`, and make `wait_for_element` take something that satisfies `Into<Selector>`?

That way they can pass in a `str` or a `Selector::XPath`. Otherwise we might have to just add separate methods for finding via XPath.

**Comment by mdrokz** [comment]
> @mdrokz nice one!
> 
> With [this](https://github.com/atroche/rust-headless-chrome/commit/b68cfa53051d625c5fb773962688cc200627b31f#diff-d91ca1f76b204ad2cf6d1781c3fbbd7559d0dd60b553dac372772308c14db558R79-R90), I think there are ways that can go wrong (e.g. if you use an XPath expression that contains a `.` or a `#` in it).
> 
> Instead, could you make strings always convert to a `Selector::CSS`, and make `wait_for_element` take something that satisfies `Into<Selector>`?
> 
> That way they can pass in a `str` or a `Selector::XPath`. Otherwise we might have to just add separate methods for finding via XPath.

I see, i will try implementing what you suggested.

**Comment by mdrokz** [comment]
> > @mdrokz nice one!
> > With [this](https://github.com/atroche/rust-headless-chrome/commit/b68cfa53051d625c5fb773962688cc200627b31f#diff-d91ca1f76b204ad2cf6d1781c3fbbd7559d0dd60b553dac372772308c14db558R79-R90), I think there are ways that can go wrong (e.g. if you use an XPath expression that contains a `.` or a `#` in it).
> > Instead, could you make strings always convert to a `Selector::CSS`, and make `wait_for_element` take something that satisfies `Into<Selector>`?
> > That way they can pass in a `str` or a `Selector::XPath`. Otherwise we might have to just add separate methods for finding via XPath.
> 
> I see, i will try implementing what you suggested.

On another thought, i think it might better to just implement seperate methods.

**Comment by mdrokz** [comment]
I made the xpath methods seperate in this commit - https://github.com/atroche/rust-headless-chrome/commit/35113dbc5837d2ebdea69d41c4e2c579a17799c2

**Comment by mdrokz** [comment]
Is this ready for merge now?
